File: js/Core/Table/CompositeColumn.js
import { unwrap, defaultArg } from "../../fable_modules/fable-library-js.4.22.0/Option.js";
import { fold, empty, singleton, collect, delay, toArray, exists } from "../../fable_modules/fable-library-js.4.22.0/Seq.js";
import { CompositeCell_$reflection, CompositeCell } from "./CompositeCell.js";
import { Record } from "../../fable_modules/fable-library-js.4.22.0/Types.js";
import { CompositeHeader_$reflection } from "./CompositeHeader.js";
import { record_type, array_type } from "../../fable_modules/fable-library-js.4.22.0/Reflection.js";

export class CompositeColumn extends Record {
    constructor(Header, Cells) {
        super();
        this.Header = Header;
        this.Cells = Cells;
    }
    static create(header, cells) {
        return new CompositeColumn(header, defaultArg(cells, []));
    }
    Validate(raiseException) {
        const this$ = this;
        return !exists((c) => !c.ValidateAgainstHeader(this$.Header, unwrap(raiseException)), this$.Cells);
    }
    TryGetColumnUnits() {
        const this$ = this;
        const arr = toArray(delay(() => collect((cell) => (cell.isUnitized ? singleton(cell.AsUnitized[1]) : empty()), this$.Cells)));
        return (arr.length === 0) ? undefined : arr;
    }
    GetDefaultEmptyCell() {
        const this$ = this;
        if (!this$.Header.IsTermColumn) {
            return CompositeCell.emptyFreeText;
        }
        else {
            let patternInput;
            const arg = [0, 0];
            patternInput = fold((tupledArg, cell) => {
                const units = tupledArg[0] | 0;
                const terms = tupledArg[1] | 0;
                if (cell.isUnitized) {
                    return [units + 1, terms];
                }
                else {
                    return [units, terms + 1];
                }
            }, [arg[0], arg[1]], this$.Cells);
            return (patternInput[1] >= patternInput[0]) ? CompositeCell.emptyTerm : CompositeCell.emptyUnitized;
        }
    }
    get IsUnique() {
        const this$ = this;
        return this$.Header.IsUnique;
    }
}

export function CompositeColumn_$reflection() {
    return record_type("ARCtrl.CompositeColumn", [], CompositeColumn, () => [["Header", CompositeHeader_$reflection()], ["Cells", array_type(CompositeCell_$reflection())]]);
}

File: src/Core/Table/CompositeColumn.fs
namespace ARCtrl

open Fable.Core

[<AttachMembers>]
type CompositeColumn = {
    Header: CompositeHeader
    Cells: ResizeArray<CompositeCell>
} with
    static member create (header: CompositeHeader, ?cells: ResizeArray<CompositeCell>) = 
        let cells = Option.defaultValue (ResizeArray()) cells
        {
            Header = header
            Cells = cells
        }

    /// Returns true if header and cells are a valid combination. E.g. Term header with term or unitized cells. IO header with freetext cells.
    ///
    /// ?raiseExeption: Default false. Set true if this function should raise an exception instead of return false.
    // TODO! Do not only check cells.Head
    member this.Validate(?raiseException: bool) =
        this.Cells
        |> Seq.exists (fun c -> c.ValidateAgainstHeader(this.Header, ?raiseException = raiseException) |> not)
        |> not

    /// <summary>
    /// Returns an array of all units found in the cells of this column. Returns None if no units are found.
    /// </summary>
    member this.TryGetColumnUnits() =
        let arr =
            [|
            for cell in this.Cells do
                if cell.isUnitized then
                    let _, unit = cell.AsUnitized
                    unit
            |]
            |> ResizeArray
        if Seq.isEmpty arr then None else Some arr

    /// <summary>
    /// Simple predictor for empty default cells.
    ///
    /// Currently uses majority vote for the column to decide cell type.
    /// </summary>
    member this.GetDefaultEmptyCell() =
            if not this.Header.IsTermColumn then
                CompositeCell.emptyFreeText
            else
                let unitCellCount, termCellCount =
                    this.Cells
                    |> Seq.fold (fun (units,terms) cell ->
                        if cell.isUnitized then (units+1,terms) else (units,terms+1)
                    ) (0,0)
                if termCellCount >= unitCellCount then
                    CompositeCell.emptyTerm
                else
                    CompositeCell.emptyUnitized

    member this.IsUnique =
        this.Header.IsUnique